### PR TITLE
[DFAJumpThreading] Rewrite the way paths are enumerated

### PR DESCRIPTION
**Pull request by UsmanNadeem**
I tried to add a limit to number of blocks visited in the paths() function but even with a very high limit the transformation coverage was being reduced.

After looking at the code it seemed that the function was trying to create paths of the form `SwitchBB...DeterminatorBB...SwitchPredecessor`. This is inefficient because a lot of nodes in those paths (nodes before DeterminatorBB) would be irrelevant to the optimization. We only care about paths of the form `DeterminatorBB_Pred DeterminatorBB...SwitchBB`. This weeds out a lot of visited nodes.

In this patch I have added a hard limit to the number of nodes visited and changed the algorithm for path calculation. Primarily I am traversing the use-def chain for the PHI nodes that define the state. If we have a hole in the use-def chain (no immediate predecessors) then I call the paths() function.

I also had to the change the select instruction unfolding code to insert redundant one input PHIs to allow the use of the use-def chain in calculating the paths.

The test suite coverage with this patch (including a limit on nodes visited) is as follows:

    Geomean diff:
      dfa-jump-threading.NumTransforms: +13.4%
      dfa-jump-threading.NumCloned: +34.1%
      dfa-jump-threading.NumPaths: -80.7%

Compile time effect vs baseline (pass enabled by default) is mostly positive: https://llvm-compile-time-tracker.com/compare.php?from=ad8705fda25f64dcfeb6264ac4d6bac36bee91ab&to=5a3af6ce7e852f0736f706b4a8663efad5bce6ea&stat=instructions:u

Change-Id: I0fba9e0f8aa079706f633089a8ccd4ecf57547ed

### CODE REVIEW

**Comment by UsmanNadeem** [comment]
> I'm busy these days. May look into the patch next week. By the way, there is compile-time regression in `ClamAV`.

I looked at the O3 config, ClamAV has 12.2% more blocks that are cloned with this patch. So that might be the reason.

**Comment by UsmanNadeem** [comment]
Ping!

**Review by XChy — COMMENTED** [review]
Sorry for late reply. Look basically good to me.

File: llvm/lib/Transforms/Scalar/DFAJumpThreading.cpp [hunk 1]
```diff
@@ -553,42 +572,118 @@ struct AllSwitchPaths {
...
+        std::vector<ThreadingPath> PredPaths =
+            getPathsFromStateDefMap(StateDef, IncomingPhi, VB);
+        for (ThreadingPath &Path : PredPaths) {
+          Path.push_back(PhiBB);
+          Res.push_back(std::move(Path));
+        }
+        continue;
+      }
+      // Not a direct prececessor, find intermediate paths to append to the
+      // existing path.
```
[the remark this answers is not in the document]
Comment (UsmanNadeem) [reply llvm/lib/Transforms/Scalar/DFAJumpThreading.cpp]: I think so. Doing it separately will save some time, not sure how much.

**Comment by UsmanNadeem** [comment]
Ping!

**Review by XChy — APPROVED** [review]
LGTM.

**Comment by mikaelholmen** [comment]
Hi @UsmanNadeem 

The following starts crashing with this patch:
```
opt "-passes=dfa-jump-threading" bbi-98332.ll -o /dev/null
```
It hits an assert like
```
opt: ../include/llvm/IR/Instructions.h:2679: llvm::Value *llvm::PHINode::getIncomingValueForBlock(const llvm::BasicBlock *) const: Assertion `Idx >= 0 && "Invalid basic block argument!"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: build-all/bin/opt -passes=dfa-jump-threading bbi-98332.ll -o /dev/null
1.	Running pass "function(dfa-jump-threading)" on module "bbi-98332.ll"
2.	Running pass "dfa-jump-threading" on function "func_17"
 #0 0x00005608ae38fc27 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (build-all/bin/opt+0x4142c27)
 #1 0x00005608ae38d70e llvm::sys::RunSignalHandlers() (build-all/bin/opt+0x414070e)
 #2 0x00005608ae39062f SignalHandler(int) Signals.cpp:0:0
 #3 0x00007f377811ccf0 __restore_rt (/lib64/libpthread.so.0+0x12cf0)
 #4 0x00007f3775cd5acf raise (/lib64/libc.so.6+0x4eacf)
 #5 0x00007f3775ca8ea5 abort (/lib64/libc.so.6+0x21ea5)
 #6 0x00007f3775ca8d79 _nl_load_domain.cold.0 (/lib64/libc.so.6+0x21d79)
 #7 0x00007f3775cce426 (/lib64/libc.so.6+0x47426)
 #8 0x00005608afa7b065 (anonymous namespace)::DFAJumpThreading::run(llvm::Function&) DFAJumpThreading.cpp:0:0
 #9 0x00005608afa73f8c llvm::DFAJumpThreadingPass::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (build-all/bin/opt+0x5826f8c)
#10 0x00005608af70d9ed llvm::detail::PassModel<llvm::Function, llvm::DFAJumpThreadingPass, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) PassBuilderPipelines.cpp:0:0
#11 0x00005608ae598cfa llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (build-all/bin/opt+0x434bcfa)
#12 0x00005608af70e41d llvm::detail::PassModel<llvm::Function, llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) PassBuilderPipelines.cpp:0:0
#13 0x00005608ae59d891 llvm::ModuleToFunctionPassAdaptor::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (build-all/bin/opt+0x4350891)
#14 0x00005608af70793d llvm::detail::PassModel<llvm::Module, llvm::ModuleToFunctionPassAdaptor, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) PassBuilderPipelines.cpp:0:0
#15 0x00005608ae59798a llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (build-all/bin/opt+0x434a98a)
#16 0x00005608af6b118b llvm::runPassPipeline(llvm::StringRef, llvm::Module&, llvm::TargetMachine*, llvm::TargetLibraryInfoImpl*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::StringRef, llvm::ArrayRef<llvm::PassPlugin>, llvm::ArrayRef<std::function<void (llvm::PassBuilder&)>>, llvm::opt_tool::OutputKind, llvm::opt_tool::VerifierKind, bool, bool, bool, bool, bool, bool, bool) (build-all/bin/opt+0x546418b)
#17 0x00005608ae357c3d optMain (build-all/bin/opt+0x410ac3d)
#18 0x00007f3775cc1d85 __libc_start_main (/lib64/libc.so.6+0x3ad85)
#19 0x00005608ae3517ae _start (build-all/bin/opt+0x41047ae)
Abort (core dumped)
```

Maybe it's the dead basic block if.then449 that trips things over?

[bbi-98332.ll.gz](https://github.com/user-attachments/files/16706107/bbi-98332.ll.gz)

**Comment by UsmanNadeem** [comment]
> Hi @UsmanNadeem
> 
> The following starts crashing with this patch:
> 
> ```
> opt "-passes=dfa-jump-threading" bbi-98332.ll -o /dev/null
> ```
> 
> It hits an assert like
> 
> ```
> opt: ../include/llvm/IR/Instructions.h:2679: llvm::Value *llvm::PHINode::getIncomingValueForBlock(const llvm::BasicBlock *) const: Assertion `Idx >= 0 && "Invalid basic block argument!"' failed.
> PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
> Stack dump:
> 0.	Program arguments: build-all/bin/opt -passes=dfa-jump-threading bbi-98332.ll -o /dev/null
> 1.	Running pass "function(dfa-jump-threading)" on module "bbi-98332.ll"
> 2.	Running pass "dfa-jump-threading" on function "func_17"
>  #0 0x00005608ae38fc27 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (build-all/bin/opt+0x4142c27)
>  #1 0x00005608ae38d70e llvm::sys::RunSignalHandlers() (build-all/bin/opt+0x414070e)
>  #2 0x00005608ae39062f SignalHandler(int) Signals.cpp:0:0
>  #3 0x00007f377811ccf0 __restore_rt (/lib64/libpthread.so.0+0x12cf0)
>  #4 0x00007f3775cd5acf raise (/lib64/libc.so.6+0x4eacf)
>  #5 0x00007f3775ca8ea5 abort (/lib64/libc.so.6+0x21ea5)
>  #6 0x00007f3775ca8d79 _nl_load_domain.cold.0 (/lib64/libc.so.6+0x21d79)
>  #7 0x00007f3775cce426 (/lib64/libc.so.6+0x47426)
>  #8 0x00005608afa7b065 (anonymous namespace)::DFAJumpThreading::run(llvm::Function&) DFAJumpThreading.cpp:0:0
>  #9 0x00005608afa73f8c llvm::DFAJumpThreadingPass::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (build-all/bin/opt+0x5826f8c)
> #10 0x00005608af70d9ed llvm::detail::PassModel<llvm::Function, llvm::DFAJumpThreadingPass, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) PassBuilderPipelines.cpp:0:0
> #11 0x00005608ae598cfa llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (build-all/bin/opt+0x434bcfa)
> #12 0x00005608af70e41d llvm::detail::PassModel<llvm::Function, llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) PassBuilderPipelines.cpp:0:0
> #13 0x00005608ae59d891 llvm::ModuleToFunctionPassAdaptor::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (build-all/bin/opt+0x4350891)
> #14 0x00005608af70793d llvm::detail::PassModel<llvm::Module, llvm::ModuleToFunctionPassAdaptor, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) PassBuilderPipelines.cpp:0:0
> #15 0x00005608ae59798a llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (build-all/bin/opt+0x434a98a)
> #16 0x00005608af6b118b llvm::runPassPipeline(llvm::StringRef, llvm::Module&, llvm::TargetMachine*, llvm::TargetLibraryInfoImpl*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::StringRef, llvm::ArrayRef<llvm::PassPlugin>, llvm::ArrayRef<std::function<void (llvm::PassBuilder&)>>, llvm::opt_tool::OutputKind, llvm::opt_tool::VerifierKind, bool, bool, bool, bool, bool, bool, bool) (build-all/bin/opt+0x546418b)
> #17 0x00005608ae357c3d optMain (build-all/bin/opt+0x410ac3d)
> #18 0x00007f3775cc1d85 __libc_start_main (/lib64/libc.so.6+0x3ad85)
> #19 0x00005608ae3517ae _start (build-all/bin/opt+0x41047ae)
> Abort (core dumped)
> ```
> 
> Maybe it's the dead basic block if.then449 that trips things over?
> 
> [bbi-98332.ll.gz](https://github.com/user-attachments/files/16706107/bbi-98332.ll.gz)

Thanks for the reproducer, I'll take a look!

**Comment by mikaelholmen** [comment]
I've seen another crash with this patch: ```opt -passes=dfa-jump-threading bbi-98705.ll -o /dev/null```
Results in:
```
opt: ../lib/Transforms/Scalar/DFAJumpThreading.cpp:332: void (anonymous namespace)::unfold(llvm::DomTreeUpdater *, llvm::LoopInfo *, (anonymous namespace)::SelectInstToUnfold, std::vector<SelectInstToUnfold> *, std::vector<BasicBlock *> *): Assertion `SI->use_empty() && "Select must be dead now"' failed.
```
@UsmanNadeem : how is debugging of these crashes (also mentioned in https://github.com/llvm/llvm-project/issues/106083) going?

[bbi-98705.ll.gz](https://github.com/user-attachments/files/16838266/bbi-98705.ll.gz)

**Comment by mshockwave** [comment]
Just want to leave a note that this patch caused performance regression (>2% increase on cycle counts and ~0.23% increase on dynamic instruction counts) in 400.perlbench for RISC-V. More specifically, after this patch DFAJumpThreading stops optimizing some cases in the hot function (i.e. S_regmatch).

**Comment by apazos** [comment]
Thanks for reporting the regression @mshockwave, let us know if the new patch from Usman resolves the issue.

**Pull request merged**